kernel/src/device/evdev/file.rs

```diff
@@ -185,6 +214,192 @@ impl EvdevFile {
+        Ok(())
+    }
+
+    /// Parses raw EVDEV ioctl command into a local variant.
+    fn parse_evdev(raw: u32) -> Option<EvdevIoctl> {
+        const IOC_NRBITS: u32 = 8;
+        const IOC_TYPEBITS: u32 = 8;
+        const IOC_SIZEBITS: u32 = 14;
+        const IOC_DIRBITS: u32 = 2;
+
+        const IOC_NRMASK: u32 = (1 << IOC_NRBITS) - 1;
+        const IOC_TYPEMASK: u32 = (1 << IOC_TYPEBITS) - 1;
+        const IOC_SIZEMASK: u32 = (1 << IOC_SIZEBITS) - 1;
+        const IOC_DIRMASK: u32 = (1 << IOC_DIRBITS) - 1;
+
+        const IOC_NRSHIFT: u32 = 0;
+        const IOC_TYPESHIFT: u32 = IOC_NRSHIFT + IOC_NRBITS;
+        const IOC_SIZESHIFT: u32 = IOC_TYPESHIFT + IOC_TYPEBITS;
+        const IOC_DIRSHIFT: u32 = IOC_SIZESHIFT + IOC_SIZEBITS;
+
+        const IOC_READ: u32 = 2;
+        const IOC_WRITE: u32 = 1;
+        const EVDEV_IOCTL_TYPE: u32 = b'E' as u32;
+        const EVIOCGNAME_NR: u32 = 0x06;
+        const EVIOCGPHYS_NR: u32 = 0x07;
+        const EVIOCGUNIQ_NR: u32 = 0x08;
+        const EVIOCGID: u32 = 0x80084502;
+        const EVIOCGVERSION: u32 = 0x80044501;
+        const EVIOCGBIT_BASE_NR: u32 = 0x20;
+        const EVIOCGKEY_NR: u32 = 0x18;
+        const EVIOCGLED_NR: u32 = 0x19;
+        const EVIOCGSW_NR: u32 = 0x1b;
+        const EVIOCSCLOCKID_NR: u32 = 0xa0;
+
+        let dir = (raw >> IOC_DIRSHIFT) & IOC_DIRMASK;
+        let type_ = (raw >> IOC_TYPESHIFT) & IOC_TYPEMASK;
+        let nr = (raw >> IOC_NRSHIFT) & IOC_NRMASK;
+        let len = (raw >> IOC_SIZESHIFT) & IOC_SIZEMASK;
+
+        if type_ != EVDEV_IOCTL_TYPE {
+            return None;
+        }
+
+        if raw == EVIOCGVERSION {
+            return Some(EvdevIoctl::GetVersion);
+        }
+        if raw == EVIOCGID {
+            return Some(EvdevIoctl::GetId);
+        }
+
+        match dir {
+            IOC_READ => match nr {
+                EVIOCGNAME_NR => Some(EvdevIoctl::GetName { len }),
+                EVIOCGPHYS_NR => Some(EvdevIoctl::GetPhys { len }),
+                EVIOCGUNIQ_NR => Some(EvdevIoctl::GetUniq { len }),
+                EVIOCGKEY_NR => Some(EvdevIoctl::GetKey { len }),
+                EVIOCGLED_NR => Some(EvdevIoctl::GetLed { len }),
+                EVIOCGSW_NR => Some(EvdevIoctl::GetSw { len }),
+                n if n >= EVIOCGBIT_BASE_NR => Some(EvdevIoctl::GetBit {
+                    event_type: n - EVIOCGBIT_BASE_NR,
+                    len,
+                }),
+                _ => None,
+            },
+            IOC_WRITE => match nr {
+                EVIOCSCLOCKID_NR => Some(EvdevIoctl::SetClockId),
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+
+    fn handle_evdev_ioctl(&self, raw: u32, arg: usize) -> Result<()> {
+        match Self::parse_evdev(raw) {
+            Some(EvdevIoctl::GetName { len }) => {
+                let evdev = self.upgrade_evdev_device()?;
+                self.write_string_to_userspace(evdev.device.name(), len as usize, arg)?;
```

Comment:
The `write_string_to_userspace` and `write_bitmap_to_userspace` methods should NOT have been made as methods of `EvdevFile`. A method accesses `self`. If it does not need `self`, then it should not be a method. Turn them into two functions.